torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1132,407 @@ struct CPUFusionFunction : public CompiledFusionFunction {
+          dim
+        );
+        drop(stack, num_inputs);
+        pack(stack, std::move(result));
+        return 0;
+      };
+    })
+});
+
+FusedKernelCache::FusedKernelCache(FusionCompiler& compiler, std::shared_ptr<Graph> _graph, int device)
+  : device(device)
+  , fallback_code(_graph)
+  , compiler(compiler)
+  , graph(std::move(_graph))
+  , input_broadcast_groups(getInputBroadcastGroups())
+  , input_chunks(getInputChunkDescriptors())
+  , kernels() {}
+
+std::atomic<size_t> FusedKernelCache::next_kernel_id {0};
+
+auto FusedKernelCache::getInputChunkDescriptors() -> std::vector<PartitionInfo> {
+  std::vector<PartitionInfo> descs;
+  descs.reserve(graph->inputs().size());
+  for (Value * input : graph->inputs()) {
+    if (Node * chunk = usedInFusedChunk(input)) {
+      descs.emplace_back(chunk->i(attr::chunks), chunk->i(attr::dim));
+    } else {
+      descs.emplace_back(1, 0);
+    }
+  }
+  return descs;
+}
 
-  auto it = cache.find(key_);
-  if (it == cache.end()) {
-    std::string name = "kernel_" + std::to_string(cache.size());
-    CompiledFusionFunction * raw_func;
-    if(agraph.device != kCPUDevice) {
-#ifdef USE_CUDA
-      raw_func = new CUDAFusionFunction(name, agraph);
-#else
-      throw std::runtime_error("cannot compile a CUDA fusion group, CUDA is not enabled.");
-#endif
+// NB: this vector is really a set, but we want to keep it contiguous in memory for faster access
+static std::vector<int64_t> getInputDependencies(Value* output) {
+  // Run a DFS traversal to find all inputs that affect a given output value
+  std::vector<Value*> queue { output };
+  std::unordered_set<Value*> inputs;
+  std::unordered_set<Value*> seen;
+  while (!queue.empty()) {
+    Value * val = queue.back(); queue.pop_back();
+    Node * producer = val->node();
+    if (producer->kind() == prim::Param) {
+      inputs.insert(val);
+      continue;
+    }
+    for (Value * input : producer->inputs()) {
+      if (/*bool inserted = */seen.insert(input).second) {
+        queue.push_back(input);
+      }
+    }
+  }
+
+  // Convert Value* into offsets into the graph's input list
+  std::vector<int64_t> offsets;
+  offsets.reserve(inputs.size());
+  for (Value * input : inputs) {
+    offsets.push_back(input->offset());
+  }
+  std::sort(offsets.begin(), offsets.end());
+  return offsets;
+}
+
+std::vector<std::vector<int64_t>> FusedKernelCache::getInputBroadcastGroups() {
+  std::unordered_set<std::vector<int64_t>, torch::hash<std::vector<int64_t>>> broadcast_groups;
+  for (Value * output : graph->outputs()) {
+    broadcast_groups.insert(getInputDependencies(output));
+  }
+  return std::vector<std::vector<int64_t>>{ broadcast_groups.begin(), broadcast_groups.end() };
+}
+
+void FusedKernelCache::run(Stack& stack) {
+  int64_t num_inputs = graph->inputs().size();
+  auto args = fmap(last(stack, num_inputs), [](const IValue& i) {
+                return i.toTensor();
```

Comment:
Yes, but I couldn't use `last`, because `ArrayRef`s only allow for const access of members, and I would need to move them back to the stack in case of a fallback. This pass can be optimized more in the future if we'll need it, but I chose to do the simplest possible thing for now.